### ISSUE TITLE
OPTIONS requests not passed on

### ISSUE THREAD

**Issue by MattSANU** [may-day/wsgicors#11]
Not all OPTIONS requests are CORS preflight requests, but the middleware currently responds to all OPTIONS requests itself; it never passes an OPTIONS request on to the next app/middleware in the chain. This prevents me from writing a later filter that responds to non-CORS OPTIONS requests with an 'Allow:' header listing the request methods I support.

Firstly, would you accept a PR to fix this issue?

Also, should the code pass on all OPTIONS requests, then decorate them with CORS headers as appropriate?
Or should it decide whether the OPTIONS request is CORS or not, and handle the request itself if it's CORS, but pass it on if it's not CORS?

I don't understand whether an HTTP request can be both a CORS preflight request *and* an attempt to discover what request methods the server actually supports. Are these two possibilities exclusive?

In other words, would it be meaningful to respond with both an Access-Control-Allow-Methods header *and* an Allow header?

**Comment by may-day** [comment]
#### 1) "but the middleware currently responds to all OPTIONS requests itself":

 oops - PR is always welcome :)

#### 2) passing on or answer exclusively in case of preflight:
On one hand if a preflight request is detected it would be sufficient if this wsgi layer answers the request. On the other hand such behaviour would contradict the wsgi stackability. But then again if no other layer answers then the typically innermost wsgi layer (usually the real application) will throw an 404 (not sure). So i'd tend to let the wsgicors layer handle a preflight request completely - your thoughts?

#### 3) Allow -header:
as i understand it, the Allow header is uri specific. So it's much more fine-grained that just the dsn:port distinction in CORS. So yes, when requesting, say, 
GET example.com/some/resource 
you could return 
Access-Control-Allow-Methods: GET, POST, PUT, DELETE
Allow: GET, POST, DELETE 

where Allow specifies the methods allowed for the resource example.com/some/resource
and Access-Control-Allow-Methods for everything below example.com

I _guess_ Allow should be a subset of or equal to the Access-Control-Allow-Methods.
Now whether this is honoured on the client-side is another question :)

**Comment by MattSANU** [comment]
Hello!

I thought about this some more. I think it makes sense for the middleware to entirely handle CORS preflight requests itself, without passing them on to the next component. Because if it had to pass on the preflight requests to some other component, so that the other component could help handle them, then the CORS middleware wouldn't be doing its job. Handling these requests is its raison d'etre, after all. So I'll send a PR to change it so that it only handles CORS preflight OPTIONS requests itself, not all OPTIONS requests. It'll be pretty simple: just check for the Access-Control-Request-Foo headers.

Regarding whether an OPTIONS request can be CORS preflight and some other sort of OPTIONS request at the same time - I don't understand the CORS specification properly on this point, but I think the two can't be combined. The preflight request is asking whether a script has permission to perform the real request - it's not actually doing a real OPTIONS request. So I think these are two conceptually separate things. However, I'd really appreciate the opinion of someone who understood the CORS specification properly, which I'm sorry I don't :-)

**Comment by MattSANU** [comment]
Ok, PR sent. This is the minimal correct change, and I believe it's wholly sufficient too, but to be honest I'm not certain of that. The Access-Control-Request-Method header is required according to the spec, but maybe there are buggy clients out there? Maybe we should check for all of the possible Access-Control-Allow-Foo headers, just to ultra-paranoid? I dunno.

**Comment by MattSANU** [comment]
Also, I think answering both Access-Control-Allow-Methods: and Allow: would be redundant/pointless. The former says to the browser "yes, you may make this request that you're asking about" - but in that case, we're talking to the web browser, not the script, so the script will never see the Allow: header. The latter says to the script "here are the headers I support" - but the browser doing a preflight doesn't care (it knows it's going to use OPTIONS), and if the script ever sees it, CORS is finished now.

Do I have this straight? I feel confused. I would appreciate someone else's opinion/guidance here.

**Comment by MattSANU** [comment]
This flowchart implies this approach is correct, but it's not a formal specification:
http://www.html5rocks.com/static/images/cors_server_flowchart.png

**Comment by may-day** [comment]
um yes, i misunderstood  "https://spring.io/understanding/CORS" as to answer also simple GETs with Access-Control-Allow-Methods - but instead it only answers with Access-Control-Allow-Origin.
So yes, it would be redundant (also nice chart).

**Comment by MattSANU** [comment]
Yeah, I can fix that too if you like. But that's benign - the extra headers don't hurt anything; they just waste a small amount of traffic. But the original problem is biting me - I want to reply to OPTIONS requests from a WSGI filter downstream of the CORS filter. So can we merge the PR please? :-)

**Comment by may-day** [comment]
done

**Comment by MattSANU** [comment]
Cheers mate. We should probably change further logic to be strictly in accordance with the spec, but that can wait til future PRs. Principle of Least Change, n'all.
